admin/server/projects.go

```diff
@@ -680,42 +681,62 @@ func (s *Server) GetProjectVariables(ctx context.Context, req *adminv1.GetProjec
+	}
+
+	resp := &adminv1.GetProjectVariablesResponse{
+		Variables:    make([]*adminv1.ProjectVariable, 0, len(vars)),
+		VariablesMap: make(map[string]string, len(vars)),
+	}
+	for _, v := range vars {
+		resp.Variables = append(resp.Variables, projectVariableToDTO(v))
+		// nolint:staticcheck // We still need to set it
+		resp.VariablesMap[v.Name] = string(v.Value)
+	}
+	return resp, nil
 }
 
 func (s *Server) UpdateProjectVariables(ctx context.Context, req *adminv1.UpdateProjectVariablesRequest) (*adminv1.UpdateProjectVariablesResponse, error) {
-	proj, err := s.admin.DB.FindProjectByName(ctx, req.OrganizationName, req.Name)
+	observability.AddRequestAttributes(ctx,
+		attribute.String("args.org", req.Organization),
+		attribute.String("args.project", req.Project),
+		attribute.String("args.environment", req.Environment),
+		attribute.StringSlice("args.variables", maps.Keys(req.Variables)),
+		attribute.StringSlice("args.unset_variables", req.UnsetVariables),
+	)
+	proj, err := s.admin.DB.FindProjectByName(ctx, req.Organization, req.Project)
 	if err != nil {
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 
 	claims := auth.GetClaims(ctx)
+	if claims.OwnerType() != auth.OwnerTypeUser {
+		return nil, status.Error(codes.PermissionDenied, "only users can update project variables")
+	}
 	if !claims.ProjectPermissions(ctx, proj.OrganizationID, proj.ID).ManageProject {
 		return nil, status.Error(codes.PermissionDenied, "does not have permission to update project variables")
 	}
 
-	proj, err = s.admin.UpdateProject(ctx, proj, &database.UpdateProjectOptions{
-		Name:                 proj.Name,
-		Description:          proj.Description,
-		Public:               proj.Public,
-		ArchiveAssetID:       proj.ArchiveAssetID,
-		GithubURL:            proj.GithubURL,
-		GithubInstallationID: proj.GithubInstallationID,
-		ProdVersion:          proj.ProdVersion,
-		ProdBranch:           proj.ProdBranch,
-		Subpath:              proj.Subpath,
-		ProdVariables:        req.Variables,
-		ProdDeploymentID:     proj.ProdDeploymentID,
-		ProdSlots:            proj.ProdSlots,
-		ProdTTLSeconds:       proj.ProdTTLSeconds,
-		Provisioner:          proj.Provisioner,
-		Annotations:          proj.Annotations,
-	})
+	if len(req.Variables) == 0 && len(req.UnsetVariables) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "no variables or unset_variables provided")
+	}
+
+	vars, err := s.admin.UpdateProjectVariables(ctx, proj, req.Environment, req.Variables, req.UnsetVariables, claims.OwnerID())
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "variables updated failed with error %s", err.Error())
+		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	return &adminv1.UpdateProjectVariablesResponse{Variables: proj.ProdVariables}, nil
+	resp := &adminv1.UpdateProjectVariablesResponse{}
+	for _, v := range vars {
+		resp.Variables = append(resp.Variables, projectVariableToDTO(v))
+	}
```

Comment:
Wouldn't the expected behavior here be that it returns all variables, not just the changed ones?

Since the endpoint semantics treats the variables as a single collection (even if the underlying database table doesn't).

Reply:
Our API resembles a patch API. I am not sure what is the best practice, whether to return all values or just what was updated. In the proposal it was specified to return the values that were created or updated by the request.

Also in order to return everything we will need to make a separate DB call which seems somewhat unnecessary.
Based on the feedback [here](https://github.com/rilldata/rill/pull/5929#discussion_r1810950851) our `UpdateProjectVariables` will become `GetProjectVariables` if user does not provide any variables to set :)

Reply:
> our UpdateProjectVariables will become GetProjectVariables if user does not provide any variables to set

Which I think aligns with the other APIs, like `UpdateOrganization` where if you send an update request that doesn't make any changes, it returns the full value (not just the changed fields), i.e. acts similar to `GetOrganization`?